routing/router.go

```diff
@@ -1668,119 +1743,209 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
-				return [32]byte{}, nil, fmt.Errorf("unable to "+
-					"route payment to destination: %v",
-					lastError)
+				return fmt.Errorf("unable to route payment "+
+					"to destination: %v", lastError)
 			}
 
-			return [32]byte{}, nil, err
+			// Terminal state, return.
+			return err
+		}
+
+		// We generate a new, unique payment ID that we will use for
+		// this HTLC.
+		paymentID, err := r.cfg.NextPaymentID()
+		if err != nil {
+			return err
+		}
+
+		attempt := &channeldb.AttemptInfo{
+			PaymentID: paymentID,
+			Route:     *route,
 		}
 
-		// Send payment attempt. It will return a final boolean
-		// indicating if more attempts are needed.
-		preimage, final, err := r.sendPaymentAttempt(
-			paySession, route, payment.PaymentHash,
+		// Before sending this HTLC to the switch, we checkpoint the
+		// fresh paymentID and route to the DB. This lets us know on
+		// startup the ID of the payment that we attempted to send,
+		// such that we can query the Switch for its whereabouts. The
+		// route is needed to handle the result when it eventually
+		// comes back.
+		err = a.NewAttempt(attempt)
+		if err != nil {
+			return err
+		}
+
+		log.Tracef("Attempting to send payment %x (pid=%v), "+
+			"using route: %v", paymentHash, paymentID,
+			newLogClosure(func() string {
+				return spew.Sdump(route)
+			}),
 		)
-		if final {
-			return preimage, route, err
+
+		// Send it to the Switch. When this method returns we assume
+		// the Switch successfully has persisted the payment attempt,
+		// such that we can resume waiting for the result after a
+		// restart.
+		err = r.cfg.SendToSwitch(route, paymentHash, paymentID)
+		if err != nil {
+			log.Errorf("Failed sending payment to switch: %v",
+				err)
+			return err
 		}
 
-		lastError = err
+		log.Debugf("Payment %x (pid=%v) successfully sent to switch",
+			paymentHash, paymentID)
+
+		return nil
 	}
-}
 
-// sendPaymentAttempt tries to send the payment via the specified route. If
-// successful, it returns the obtained preimage. If an error occurs, the last
-// bool parameter indicates whether this is a final outcome or more attempts
-// should be made.
-func (r *ChannelRouter) sendPaymentAttempt(paySession *paymentSession,
-	route *Route, paymentHash [32]byte) ([32]byte, bool, error) {
+	switch {
 
-	log.Tracef("Attempting to send payment %x, using route: %v",
-		paymentHash, newLogClosure(func() string {
-			return spew.Sdump(route)
-		}),
-	)
+	// If this payment had no existing payment ID, we make a new attempt.
+	case a.Attempt == nil:
+		if err := sendNewAttempt(); err != nil {
+			return [32]byte{}, nil, err
+		}
 
-	preimage, err := r.sendToSwitch(route, paymentHash)
-	if err == nil {
-		return preimage, true, nil
+	// Otherwise we'll check if there's a result available for the already
+	// existing payment ID.
+	default:
 	}
 
-	log.Errorf("Attempt to send payment %x failed: %v",
-		paymentHash, err)
+	for {
+		// We'll ask the switch whether this is a known paymentID.
```

Comment:
I still find the `sendNewAttempt` closure not the easiest way to read the code. Basically this for loop can always start with executing everything inside `sendNewAttempt`. Then in the cases below, it doesn't need to be called anymore before the `continue`. The only exception that needs to be made is that it should be skipped the first time for a resumed payment. Maybe with a boolean `resuming` that is initialized outside the loop with `true` if it is a resumed payment. If - at the top of the loop block - `resuming` is true, the `sendNewAttempt` logic will be skipped and the `resuming` bool will be set to false.

Reply:
Agreed. Did this by using `attempt == nil` to indicate a new attempt should be made :)